Move the Block Patterns UI to the inserter

youknowriad opened the pull request:
Refs #17335

This PR is the second iteration on the block patterns UI. It moves the block patterns into the inserter.

<img width="314" alt="Capture d’écran 2020-04-10 à 10 49 05 AM" src="https://user-images.githubusercontent.com/272444/78981929-ec2c9e00-7b18-11ea-86b0-0a9a067e5765.png">

It is only enabled for root-level insertions.

Potential follow-up tasks:

 - Support searching in the block patterns tab.
 - Disable patterns if one of the blocks is not allowed.
 - Allow inserting patterns inside container blocks that support it.

youknowriad committed:
Move the Block Patterns UI to the inserter

youknowriad commented on packages/block-editor/src/components/inserter/block-list.js:
```diff
@@ -3,7 +3,6 @@
  */
 import {
 	map,
-	pick,
```
The ode removed in this file is moved up to be shared with the patterns.

mtias commented:
I don't think we should do this until we have a finalized design for the inserter as a whole.

youknowriad commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -99,6 +117,17 @@ $block-inserter-search-height: 38px;
 
+// This extra div is needed because
+// flex grow and overflow auto doesn't work well together.
+.block-editor-inserter__scrollable {
```
I spent hours trying to make scrolling behave properly without this extra div without any luck. There's a mix between flex, auto-height popover computation, and scrolling that makes it hard to achieve. This solution is decent I think.

enriquesanchez commented:
>Support searching in the block patterns tab

We had envisioned that the top search field will act as a global search, for both blocks and block patterns. Something like this:

<img width="638" alt="Screen Shot 2020-03-17 at 15 19 16" src="https://user-images.githubusercontent.com/3276087/76903000-bcb0ab80-6862-11ea-9248-b3284f2a27a2.png">

Would that be technically feasible? 

>Disable patterns if one of the blocks is not allowed

Should disabled patterns not show up at all instead? I wonder about how to communicate why a pattern is disabled.

>Allow inserting patterns inside container blocks that support it.

Do you mean containsers such as the group block? Are there any others you had in mind?

youknowriad commented:
> Would that be technically feasible?

Yes, it's possible, the idea is to build things in steps as this is mentioned as a follow-up.

> Should disabled patterns not show up at all instead? I wonder about how to communicate why a pattern is disabled.

When I say "disable", it doesn't necessarily mean a disabled button, it can be hidden.

> Do you mean containsers such as the group block? Are there any others you had in mind?

Group, columns,  cover,  media & text and any third-party container block.

--- 
@mtias @enriquesanchez Let me know when you have a good design you want me to use as a target.

mtias commented:
> Allow inserting patterns inside container blocks that support it.

I've been thinking a bit about how this could be handled, particularly for template building / theme editor. Perhaps we need a `core/pattern` block that is basically just a placeholder to render pattern miniatures, or to force a pattern-only inserter to be used. (This would be similar to the "section" block discussed a long time ago.) I'd expect a common way to build templates would be to add and stitch together patterns.

jasmussen commented:
About this branch: this feels very promising, because I believe it's important that there remains only a single unified button to insert content. However I agree with Matías, it would be good to have a design ready before we go here. 

One observation: clicking the Patterns tab takes a few milliseconds longer, presumably because it loads the block preview component that renders each block. This is with only 3 patterns present. It seems worth to:

a) try this with a bunch more patterns to see how much worse it gets, even if it's just duplicates of the same patterns
b) consider some way to asynchronously load each pattern, ie. a spinner inside each preview container so nothing is blocked as things load. 

> I've been thinking a bit about how this could be handled, particularly for template building / theme editor. Perhaps we need a core/pattern block that is basically just a placeholder to render pattern miniatures, or to force a pattern-only inserter to be used. (This would be similar to the "section" block discussed a long time ago.) I'd expect a common way to build templates would be to add and stitch together patterns.

The placeholder component is a powerful pattern for surfacing complex features in a simple-to-use way. In that vein, it seems like it could be a good way to surface patterns, which might otherwise "feel complex" when in fact it's supposed to be the opposite. If a section of the page shows thumbnails for you to pick from, it can be as little as one click to get something nice going.

But it does raise a few questions:

- While we can make the placeholder element-query responsive now (and presumably collapse to a button that opens a modal) — would we want to limit such a pattern block to only the top level? Or an addon-question: how much sense does it make to insert a pattern as a nested block?
- Patterns are intrinsically _sections of a page_, not a full page. Which means in order to build a whole page with just patterns, you need to insert several in succession. 
  - Does that mean we insert multiple pattern placeholders in sequence to suggest this, and what would that look like?
  - Does it mean another placeholder gets inserted or suggested once you've inserted one pattern?

A placeholder for patterns really seems to make the most sense when inserted at the root level. In that vein, perhaps to truly enable page building with patterns, this suggests a rethink of the "trailing appender". This one:

<img width="719" alt="Screenshot 2020-03-23 at 08 48 32" src="https://user-images.githubusercontent.com/1204802/77293784-505efd80-6ce3-11ea-99fd-b2a1c5b81746.png">

What if this interface defaulted and heavily leaned into suggesting patterns as the next section of the page. What would that look like? Would it default to inserting pattern placeholders, or would it simply show an interface to browse the patterns there?

youknowriad commented:
> clicking the Patterns tab takes a few milliseconds longer, presumably because it loads the block preview component that renders each block. This is with only 3 patterns present

This is an issue regardless of the UI, we definitely need to find ways to improve this (even in the sidebar). I believe @retrofox @marekhrabe @obenland and friends already faced and explored solutions for this?

youknowriad commented:
> how much sense does it make to insert a pattern as a nested block?

Inserting patterns inside blocks definitely make sense for me. Groups, columns, even post-content blocks

jasmussen commented:
> Groups, columns, even post-content blocks

To me those aren't patterns. Those are blocks.

I'm eternally open to be disagreed with and overruled here, but to me a patterns is _multiple blocks that are customized and visually styled_. Columns is an ingredient in that, not a pattern itself.

youknowriad commented:
@jasmussen Looking at the current patterns, I don't see why I won't be able to insert the "two columns of text" pattern inside a group block and apply some tweaks (background color, full wide,...) to the container

jasmussen commented:
> Looking at the current patterns, I don't see why I won't be able to insert the "two columns of text" pattern inside a group block and apply some tweaks (background color, full wide,...) to the container

I don't disagree. I'm personally fine if we think patterns can be inserted at any nesting level. In the case of limiting the nesting level, I was primarily referring the theoretical "pattern block" that Matías suggested.

mapk commented:
Here's a couple of my thoughts on this as well...

I also view patterns as a section of blocks. For me, this means a couple of things. 

1. When a pattern is added, I view it as a single unit since I added it as a unit. For this reason, I should be able to select the pattern and delete the whole thing as a single unit. This leads me to think every pattern should be in a container (ie. Group block).
2. Because I've always imagined patterns as a "section of a page," I always thought they'd be more than just two columns of text, or two buttons side-by-side. Those are really just blocks that provide patterns. 😕  I imagined them as a complete section of a page – most likely something more complex. In this scenario patterns wouldn't be able to be nested inside other blocks from the gitgo. Now if a user wanted to drag it into another block once it's been added to the page, I could see allowing that, but not adding it directly inside a block. The Section block mentioned above is an exception.
3. If patterns are now becoming something much more simplified than actual page sections (ie. two columns of text or two buttons side-by-side) then we should allow them to be added inside other blocks.
4. I do prefer the ability to preview patterns from the Inserter and drop them in without a Pattern block first. I haven't tested this PR, but how does it differ from https://wordpress.org/plugins/design/ ? These seem to load quite speedily.

I'm open to any disagreements here. 😄 These notes are just my paradigm of patterns.

Do we need to define what patterns are further? Maybe something like, _a Pattern must include 2 or more blocks_.

retrofox commented:
> > clicking the Patterns tab takes a few milliseconds longer, presumably because it loads the block preview component that renders each block. This is with only 3 patterns present
> 
> This is an issue regardless of the UI, we definitely need to find ways to improve this (even in the sidebar). I believe @retrofox @marekhrabe @obenland and friends already faced and explored solutions for this?

Yeah, definitely something to improve. I haven't taken a look yet to this PR, but pretty sure it happens in other places where the `<BlockPreview />` is used.

obenland commented:
> [Previews taking a long time to load] is an issue regardless of the UI, we definitely need to find ways to improve this (even in the sidebar). I believe @retrofox @marekhrabe @obenland and friends already faced and explored solutions for this?

We faced it with the page layout selector on wp.com and worked around it by using static images instead. [We traced it down to `setState` taking a long time](https://github.com/Automattic/wp-calypso/pull/35713#issuecomment-536676899) but couldn't find a work around.

mtias commented:
For the display of pattern previews we should apply a [windowing technique](https://react-window.now.sh/#/examples/list/variable-size).

youknowriad commented:
Based on the last designes on https://github.com/WordPress/gutenberg/issues/21080 It seems like this PR is moving toward the right direction.

For the sake of iteration, I'd like to propose that we move forward with the current state and continue the inserter iterations on small PRs to achieve the designs there. I believe the proposal is good enough and better than master (which mean mergeable state). 

Unless you think we should do it all in one PR?

youknowriad committed:
Restore the patterns panel gray background

mtias commented:
> Unless you think we should do it all in one PR?

I wouldn't want to release many interim stages in the plugin because inserting blocks is a core part of the experience.

youknowriad committed:
Redesign search form and tabs

youknowriad committed:
Redesign the inserter panels and resize it

youknowriad committed:
Redesign the preview panel

youknowriad committed:
Try using a panel instead of a dropdown

mtias commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -55,22 +183,29 @@ function InserterMenu( {
+							},
+							{
+								name: 'patterns',
+								title: __( 'Block Patterns' ),
```
Let's switch this to just Patterns for now.

youknowriad committed:
Use a popover for the block preview

youknowriad committed:
User a border to separate the tips from the rest

youknowriad committed:
Rename block patterns to just patterns

mtias commented:
A few notes:

- The container should be 350px in width.
- Opening the right hand sidebar (either through the cog menu or a plugin) should probably close the inserter panel, at least below a certain viewport threshold (< 1300).
- When the inserter is opened, the + button in the header should look activated (dark background instead of blue).
- Let's try the 12px font size for the block names as per the design to see how it fares in context. (It should make buttons look more similar to the PR that seeks to add labels to the header icons.)
- We should definitely allow patterns to be inserted inside other containers (right now the patterns tab disappears when you are inside a Cover block or other nested structure).
- Can we restore the preview on blocks?
- Every time you switch tabs the patterns get re-rendered seemingly causing the async loading to re-trigger.
- The border beneath the tabs should be the same lightness as the sidebar one. It's currently darker.
- Not related to this PR, but we have lost the blue line indicating where blocks will be inserted.
- We need a stronger focus outline on the search input, and white background when focused.

![image](https://user-images.githubusercontent.com/548849/78983750-184e1c00-7b25-11ea-8b9d-c7ada268aac1.png)

youknowriad committed:
design tweaks

youknowriad committed:
update the search focus style

youknowriad committed:
better spacing

youknowriad committed:
Mutually exclusive sidebars

mtias commented:
@youknowriad do you think we should make the open inserter a region now?

mtias commented:
Looks like we lost the added spacing between the inserter and the next tool here:

![image](https://user-images.githubusercontent.com/548849/78989395-ce206700-7b33-11ea-985d-199c662df7d7.png)

youknowriad commented:
> @youknowriad do you think we should make the open inserter a region now?

Yes, it's already the case in the PR

youknowriad committed:
Fix the inserter toggle

youknowriad committed:
Remove duplicate region

mtias commented:
For the search, I think we should replace the magnifying glass icon and show the X from the icon set when there's a value in the input. Also probably disable the auto-complete:

![image](https://user-images.githubusercontent.com/548849/79114151-dc17f700-7d82-11ea-9925-d81c82ccadd0.png)

jasmussen commented:
Some things about this I really, really like. 

- By moving in the canvas and not covering anything, it will be so much more obvious where blocks will be inserted, especially once #21475 gets addressed, which I believe @ellatrix has a fix to.
- The whole thing feels spacious and accommodating, and a good place to hold both patterns and blocks. It's a literally first class experience.
- Blocks & their icons look neater and tidier than ever.
- Block patterns load so nicely! No longer that stutter waiting for them all to instantiate — you made some async magic work here, and it's glorious!

---

Some small things:

The vertical centering of the preview window, as opposed to at the top or in the axis of the hovered block feels slightly off:

![preview](https://user-images.githubusercontent.com/1204802/79115298-c22be380-7d85-11ea-9bcb-255013bfcaff.gif)

When block patterns load, the scrollbar appears after a while, pushing things inwards. This is fine to ignore as it's mostly a non-issue. But I wonder if we should just force `overflow-y: scroll;` here to refine that just a smidge.

Speaking of block patterns, they are bordered and on a gray background, but in #21080 they are not bordered and on a white background:

<img width="397" alt="Screenshot 2020-04-13 at 12 45 56" src="https://user-images.githubusercontent.com/1204802/79115511-4e3e0b00-7d86-11ea-8484-207bbef3113e.png">

I'd love to get us closer to #21080, as that make it feel more like contents in the white popover.

<img width="287" alt="Screenshot 2020-04-13 at 12 43 39" src="https://user-images.githubusercontent.com/1204802/79115595-8cd3c580-7d86-11ea-83c8-9bf246348c5f.png">

Do we need the "Video" title duplicated in the preview? Maybe. If yes, then we should make it the same 13px font size as the preview text itself. 

We should also change "No Preview Available." to "No preview available".

---

Some things to think about, possibly bigger:

When you have the inserter open and select a block, the inserter is filtered to child blocks of that block. This isn't super obvious — the blue line will very probably help a lot here. But it gets somewhat muddier when you select a block that doesn't allow any chlid blocks, i.e. a Columns block (compared to a _Column_ block):

![filter](https://user-images.githubusercontent.com/1204802/79115712-dc19f600-7d86-11ea-8f91-8bbeb91baf84.gif)

This is probably not something to fix in this branch, as it is in fact existing behavior. But it is something to think about. A solution could be to show the filter that is being applied ("Blocks allowed inside Columns") somewhere, to at least indicate why the contents of the inserter suddenly changed. 

Finally — Right now if you have the sidebar open, and open the inserter, the sidebar closes, and vice versa. This feels necessary for many breakpoints, but not all of them. On my spacious monitor, it does not feel necessary, and I'd like to have both. I wonder if we should only close one or the other if the screen is <1600px?

ellatrix commented:
I'm testing a bit on mobile. There's no way to close the modal/sidebar?

jasmussen commented:
To Matías point:

> Opening the right hand sidebar (either through the cog menu or a plugin) should probably close the inserter panel, at least below a certain viewport threshold (< 1300).

We have these two existing breakpoints one of which might be nice to reuse here:

$break-huge: 1440px;
$break-wide: 1280px;

enriquesanchez commented:
After submitting a search query, tabbing does not move focus to the list of results. It moves back up to the "W" logo in the top toolbar instead.

![2020-04-13 16-07-38 2020-04-13 16_08_43 2020-04-13 16_10_032](https://user-images.githubusercontent.com/3276087/79161565-50619300-7da1-11ea-846b-dd941fa74a9f.gif)

enriquesanchez commented:
I think that having an aways-visible panel/sidebar for the block inserter is a great way to encourage creation and discoverability, but one thing I'm finding odd is how it interacts with the _other sidebar_. 

Having the right sidebar close automatically feels unexpected, and if you're someone who relies on the block inspector a lot, having to reopen it every time you interact with the inserter feels annoying.

I understand the reason why currently this has to happen, but I don't think that we should sacrifice one thing for the other. 

@MichaelArestad [suggested an interesting workaround](https://github.com/WordPress/gutenberg/issues/21080#issuecomment-610955737):

>I would love a persistent sidebar when I'm on a device with a large screen that lets me insert content when I wish. [...] I also see the need for an overlay that is temporary for smaller screen widths similar to how the top inserter currently functions.

Would that be possible? I think this gives us some middle ground that does not sacrifice UX.

mapk commented:
So good!  The design feedback I have is similar to others above.

> The vertical centering of the preview window, as opposed to at the top or in the axis of the hovered block feels slightly off:

I tend to agree with this and would like to see the preview along the same horizontal axis as the block I'm hovering.

> Speaking of block patterns, they are bordered and on a gray background, but in #21080 they are not bordered and on a white background: I'd love to get us closer to #21080, as that make it feel more like contents in the white popover.

I'm hesitant about this. When the patterns don't have any borders or background colors, it is difficult to understand when one pattern ends and another begins without having to hover over it. @jasmussen, do you feel the hover effect would be enough here? That being said, It would be great to allow full wide patterns to reach the edges of the Inserter to convey this. 

>But it gets somewhat muddier when you select a block that doesn't allow any chlid blocks, i.e. a Columns block (compared to a Column block):

@jasmussen, there is an issue for that one. :) https://github.com/WordPress/gutenberg/issues/19354

>Having the right sidebar close automatically feels unexpected, and if you're someone who relies on the block inspector a lot, having to reopen it every time you interact with the inserter feels annoying.

This makes sense as well. Having to reopen the sidebar every time I use the Inserter feels like a bug.

### Additional feedback

The blue block categories looked like links to me. I tried clicking on them multiple times. We should revisit the color see as it matches the color of the "Save Draft" link in the top toolbar.

![Screen Shot 2020-04-13 at 4 00 57 PM](https://user-images.githubusercontent.com/617986/79169026-29ef2800-7da0-11ea-91ab-7db08582a1d9.png)

We need an accessibility check on this search box, @enriquesanchez. With no border and a very subtle grey background, I don't think this is accessible.

![Screen Shot 2020-04-13 at 4 02 51 PM](https://user-images.githubusercontent.com/617986/79169082-5acf5d00-7da0-11ea-9a02-1bb6d4fbbb5d.png)

I'm really excited to see this progress! Excellent work, y'all. 😄

jasmussen commented:
> I'm hesitant about this. When the patterns don't have any borders or background colors, it is difficult to understand when one pattern ends and another begins without having to hover over it. 

I hear you, but I also feel like the editor has suffered visually by us doing "defensive design" over the years; poring and discussing over static mockups until they check every mental box, instead of realizing the truth: it's 4 lines of CSS code and we can ship it in a PR and try it out and see if that mental checkbox is an actual problem or not.

To phrase it differently: the mockup Pablo made looks so completely beautiful and bold to me without those borders, so beautiful that I think it's really important we actually try it, before we theorize it to not work. Especially as it's so easy to revisit. 

And yes, it needs the same solid focus outline that everything else has! Which I expect we'll find to be plenty.

youknowriad committed:
Use a fixed position for the block preview

youknowriad commented:
Unfortunately, the Popover component is not smart enough to handle positioning the preview relatively to the block row, category... I've updated the PR to use absolute positioning to avoid blocking this PR because of this. Support this might take time and is better handled on its own PR. Related #21275

youknowriad committed:
Disable search autocomplete

youknowriad committed:
Style tweaks

youknowriad commented:
I've tried the white patterns, let me know how it feels and applied some suggested tweaks.

youknowriad commented:
> When you have the inserter open and select a block, the inserter is filtered to child blocks of that block. This isn't super obvious

For me, we should only show the list of the child blocks without any categories in this case too, what do you think?

youknowriad committed:
Disable auto closing sidebars on large screens

youknowriad commented:
> I'm also not able to directlt move focus to the list of blocks after opening the inserter.

What do you mean here? When I use "Tab" it does move the focus to the blocks.

youknowriad commented:
@enriquesanchez using a dropdown for small screens and a sidebar for large ones is possible I think. It could have some challenges when resizing the screen. If it's a direction we want to go to, I'd be happy to try it but I think it should be its own PR.

jasmussen commented:
> I've tried the white patterns, let me know how it feels and applied some suggested tweaks.

This is what I see:

<img width="399" alt="Screenshot 2020-04-14 at 14 00 26" src="https://user-images.githubusercontent.com/1204802/79222703-51a1c700-7e58-11ea-9adf-14ee0c860847.png">

That's cool, but it's still a lot of borders and paddings, and it's important here to not just look at each individual pattern in isolation, but to look at the whole editor holistically, which has been very very border heavy for a long time. What ever we can do to get closer to this, would be good. 

<img src="https://user-images.githubusercontent.com/25516611/77783526-63e5cd80-702f-11ea-9876-2a127b1ffc15.png">

In using the inspector, I could get pretty close to this by removing borders and paddings. Some paddings appear hard-coded into the preview — for ever pixel we remove here, it also results in a bigger preview.

jasmussen commented:
> For me, we should only show the list of the child blocks without any categories in this case too, what do you think?

Yes if we can remove the categories that would benefit a lot, seeing this three times isn't helpful:

<img width="349" alt="Screenshot 2020-04-14 at 14 07 20" src="https://user-images.githubusercontent.com/1204802/79223274-49965700-7e59-11ea-811b-b7df9b07fc63.png">

youknowriad commented:
@jasmussen so I removed borders, padding and pattern titles too?

jasmussen commented:
>  so I remove borders, padding and pattern titles too?

Yes. If you feel uncomfortable, I can try this in a separate PR so as to not keep this up. 

It's always going to be a back and forth when implementing mockups. When I implement something, I try to do it 100% faithfully to the design, so as to capture the essence. It doesn't always work in practice, and then we revisit.

youknowriad committed:
Simplify the patterns design further

youknowriad commented:
Removed the borders, paddings and title.

youknowriad committed:
Only show the child blocks once

youknowriad committed:
Add an aria label to the patterns

enriquesanchez commented:
>What do you mean here? When I use "Tab" it does move the focus to the blocks.

@youknowriad Ah, I think I was confused by the two "invisible" tab stops that happen between the search box and the first block. I assume it's invisibly focusing on the tabs? Maybe it's missing focus styles?

![2020-04-14 18-10-10 2020-04-14 18_10_50](https://user-images.githubusercontent.com/3276087/79282721-883d0900-7e7b-11ea-92ce-18b9414a00ec.gif)

youknowriad committed:
Try adding focus styles

youknowriad commented:
I'm working on fixing the tests but how is this looking for a v1? How far are we from "merge and iterate"?

youknowriad committed:
Fix unit tests

pablohoneyhoney commented:
Nice work @youknowriad, as usual. Here a few notes based on what I'm testing:

- Smooth the panel sliding in.
- I agree with @jasmussen hiding labels in search results.
- Radius of selection states for both blocks and patterns to be 2px, not as rounded. 
- To @ellatrix point, on mobile viewport the top bar disappears so I can't close the inserter. I'm unsure if a dropdown adds more value that a flyover as you'd likely don't see much content anyways, and users likely won't create dense pages/posts on the move. But at least I should be able to close it. 

**Additionally (I'll follow up with a mock on)**
- Show Resusable blocks as a third tab.
- Move the tip to show in Search context.
- Refine Search field behavior.

**Needs further thinking but**
We need to add a visual signal in the canvas to understand where the block or pattern is going to land. 
I wonder how useful is to view the breadcrumb when the inserter is open. It's more distracting than anything.

pablohoneyhoney commented:
Re: **Refine Search field behavior.**

**Inactive**. It says only _Search_ to be inclusive of blocks and patterns. 

<img width="515" alt="Screen Shot 2020-04-15 at 12 28 12 PM" src="https://user-images.githubusercontent.com/25516611/79362533-b4f12f00-7f14-11ea-86b0-118dcc9a0284.png">

**Active** Search fades to grey. _icon-search_ becomes _icon-close-small_

<img width="552" alt="Screen Shot 2020-04-15 at 12 28 21 PM" src="https://user-images.githubusercontent.com/25516611/79362559-bde20080-7f14-11ea-9496-c34e1167bfda.png">

**Typing.** Labels are even less relevant in this context. 


<img width="561" alt="Screen Shot 2020-04-15 at 12 28 33 PM" src="https://user-images.githubusercontent.com/25516611/79362840-13b6a880-7f15-11ea-96f4-8558c005af22.png">

**Hover**. Once the user engages with a block, the canvas highlights its position.

<img width="1027" alt="Screen Shot 2020-04-15 at 12 28 51 PM" src="https://user-images.githubusercontent.com/25516611/79362908-2c26c300-7f15-11ea-83ea-789871062684.png">

shaunandrews commented:
![hover styles](https://user-images.githubusercontent.com/191598/79359673-e23bde00-7f10-11ea-9954-2dfe04669f1b.gif)

The hover styles for blocks and patterns are currently different; Blocks use a single pixel, blue border while patterns are using a thicker black border. I'd expect them to be the same.

Also in the GIF above you'll see the `:focus` state for the tabs; I'm not a fan. Maybe give a border radius?

--

![strangeness when clicking](https://user-images.githubusercontent.com/191598/79362532-b4f12f00-7f14-11ea-8e63-189bd39c9bd0.gif)

If I select a Column the inserter goes blank. I'd expect to be able to put blocks and patterns in the column. The logic behind what the inserter shows and when is not obvious right now.

pablohoneyhoney commented:
A follow up with the No results state.

Worth noting that the tip piece might need its own PR and further work, but it's relevant in this context. 

<img width="673" alt="Screen Shot 2020-04-15 at 12 54 28 PM" src="https://user-images.githubusercontent.com/25516611/79365209-4f9f3d00-7f18-11ea-8d34-432868f786d3.png">

shaunandrews commented:
<img width="703" alt="image" src="https://user-images.githubusercontent.com/191598/79371892-17e9c280-7f23-11ea-9c8a-47a3f69745a1.png">

I also noticed that the examples use theme styles, which can sometimes clash with the grey background.

youknowriad committed:
Fix e2e tests

mtias commented:
@youknowriad if there are concerns with collapsing the main sidebar let's split that out so we can refine that interaction separately and get this one merged.

youknowriad committed:
Animate the inserter

youknowriad committed:
border radius fixes

youknowriad committed:
Add close button on mobile

mtias commented:
Clicking on the panel has a weird focus outline:

![image](https://user-images.githubusercontent.com/548849/79488109-86a94780-8019-11ea-80de-0c30be22a804.png)

mtias commented:
I think when the panel is open, the + icon should not have the blue color on hover as it makes it seem as it hasn't changed at all.

mtias commented on packages/interface/src/components/interface-skeleton/index.js:
```diff
@@ -38,6 +39,7 @@ function InterfaceSkeleton( {
 	const defaultLabels = {
 		header: __( 'Header' ),
 		body: __( 'Content' ),
+		leftSidebar: __( 'Left sidebar' ),
```
Shouldn't we call this "Block Library" or "Insert Blocks"?

youknowriad replied:
I guess so, this is a generic UI component though, so I'll pass a labels prop.

mtias commented on packages/interface/src/components/interface-skeleton/index.js:
```diff
@@ -64,6 +66,17 @@ function InterfaceSkeleton( {
+					<div
+						className="interface-interface-skeleton__left-sidebar"
+						role="region"
+						/* translators: accessibility text for the settings landmark region. */
```
The translation text seems wrong

mtias commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -112,6 +121,18 @@ function Layout() {
 			hasBlockSelected ? 'edit-post/block' : 'edit-post/document'
 		);
 
+	// Inserter and Sidebars are mutually exclusive
```
Should we leave this one out for a separate PR?

mtias commented on packages/components/src/tab-panel/style.scss:
```diff
@@ -32,6 +37,15 @@
 
 }
+
+.components-tab-panel__tab-content:focus {
```
Why do we need this?

youknowriad replied:
It's related to the comment above from @enriquesanchez basically, if you tab in the inserter, you'll have two tab stops without focus styles if we remove the changes on this component. I think the focus styles I added are not great and I'd love some designer CSS help to improve them.

jasmussen replied:
Can we remove the tab stop that here entirely?

I'll see if I can help with the active tab focus.

youknowriad replied:
I don't think we can, at least the component (TabPanel) had very extensive a11y testing when it was built. I'd defer to others about alternatives though.

jasmussen replied:
Understood. Only asking because it used to be able to collapse but doesn't anymore, and the document sidebar has tabs but doesn't have a stop for "tab contents".

youknowriad replied:
We're not using the TabPanel component there and just a list of buttons. I can update if needed but I do believe the TabPanel had more a11y thinking in it. @diegohaz Do you have thoughts on this? (the best way to achieve tabs, can we remove the "tab content" tab stop?)

diegohaz replied:
It would be good to have some real user testing for this particular widget. But the WAI-ARIA recommendations say this:

> Tab: When focus moves into the tab list, places focus on the active tab element. When the tab list contains the focus, moves focus to the next element in the page tab sequence outside the tablist, **which is typically either the first focusable element inside the tab panel or the tab panel itself**.
>
> — https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-19

So, in theory, as long as you have a focusable element inside the tab panel, you don't need to make the container focusable.

`role="tabpanel"` elements usually have an `aria-labelledby` attribute pointing to their corresponding `role="tab"` element. Whenever you enter a tab panel with NVDA, it will announce it and its label, regardless of whether you're focusing the tab panel itself or an item inside it.

If the element has `role="tabpanel"`, VoiceOver will always move to the tabpanel when using <kbd>Control</kbd> + <kbd>Option</kbd> + Arrow keys. In this case, making it focusable or not doesn't really matter.

youknowriad replied:
Great thanks for the details here @diegohaz I believe we can try without tab stop then and test 👍

mtias commented on packages/components/src/button/style.scss:
```diff
@@ -249,6 +249,19 @@
 	}
 
+	// Toggled style.
+	&.is-pressed {
```
We should handle hover.

mtias commented on packages/block-editor/src/components/inserter/panel.js:
```diff
@@ -0,0 +1,22 @@
+
+function InserterPanel( { title, icon, children } ) {
+	return (
+		<div>
```
Is this div necessary?

mtias commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -55,25 +183,34 @@ function InserterMenu( {
+							},
+							{
+								name: 'patterns',
+								title: __( 'Patterns' ),
```
Should have a translator comment clarifying it's "block or design patterns"

mtias commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -124,7 +145,53 @@ function Layout() {
+					leftSidebar={
+						mode === 'visual' &&
+						isInserterOpen && (
+							<Animate
```
I think the animation is a bit off. Can we remove it a add it back with springs?

youknowriad replied:
What's wrong with the animation, it's the same one used in the publish panel.

mtias replied:
The easing. Neither seems refined enough.

youknowriad committed:
remove tab stop from the inserter

youknowriad committed:
Rename the left sidebar label

youknowriad committed:
fix translator comments

youknowriad committed:
fix isPressed buttons hover styles

youknowriad committed:
More translator comments for the inserter tabs

youknowriad committed:
Remove the animation and fix e2e test

youknowriad committed:
Remove useless div

youknowriad commented:
@mtias Addressed the feedback, it should be better now.
There still work needed based on the feedback above from @pablohoneyhoney @shaunandrews and @jasmussen but I believe we're on a good shape to land and iterate.

youknowriad committed:
Improve preview

jasmussen commented:
Nice work Riad, thanks for staying on this.

- Looks great
- Feels good overall, including tabbing

I'd like to polish the tab focus and the appearance of patterns more separately, also the tip. But doesn't have to be part of this one.

A few small things. The focus style for blocks is cropped:

<img width="147" alt="Screenshot 2020-04-17 at 11 55 24" src="https://user-images.githubusercontent.com/1204802/79557389-c36d5100-80a2-11ea-9749-341bb5bc3b28.png">

There's overflow on the container, an inner padding of a pixel is probably sufficient to fix it. 

The patterns have an invisible extra tab-stop:

![invisible tab stop](https://user-images.githubusercontent.com/1204802/79557435-d54ef400-80a2-11ea-944e-afd79786ce0b.gif)

Not quite sure what captures that focus, perhaps the iframe preview inside?

The final thing is the behavior of collapsing the sidebar. As [Matías suggested](https://github.com/WordPress/gutenberg/pull/20951#issuecomment-614585980), we could omit that for now to get this in and iterate. Alternately, I'd say we reduce the breakpoint for this behavior happening quite a bit. Right now it appears that the sidebar-closing happens < 1440 ($break-huge). I think we can reduce that to $break-large (1280) and it will happen in much fewer instances, and then we can tweak that further as we learn from it in master.

youknowriad committed:
Center blocks by default

youknowriad commented:
@jasmussen I was not able to find a good fix for the "cropped thing", in fact I don't understand the issue yet :P can you help?

The patterns tab stop, yes, it's the iframe. I though @aduth fixed a related issue on the third-party library used there but maybe it's not updated yet?

youknowriad committed:
Smaller breakpoint for the mutually exclusive sidebar closing

youknowriad committed:
tweak tab styles

mtias reviewed (approved):
Let's get this in and iterate, thanks for addressing all the issues raised so far.

jasmussen commented:
Actually before you merge, I'm seeing one regression. The normal block toolbar is-pressed state appears affected negatively by changes made in this PR:

![hover](https://user-images.githubusercontent.com/1204802/79567581-c7a26a00-80b4-11ea-8f92-25c729346def.gif)

Here's how it should look:

![should](https://user-images.githubusercontent.com/1204802/79567625-dab53a00-80b4-11ea-971b-38eac7a0a3f6.gif)

youknowriad committed:
Fix toolbar hover styles

youknowriad commented:
Going to revert the change about the auto-closing breakpoint. Our tests break spectacularly because of that, it seems there's no easy way to change the browser window size in Puppeteer at the moment so we're always considered in a big screen.

youknowriad committed:
Revert "Smaller breakpoint for the mutually exclusive sidebar closing"

This reverts commit ea6a3445b0d55704de24997c9e0859eb771bbd03.

simison commented:
One thing to double-check is that it's not possible to render empty patterns menu by unregistering all patterns (https://github.com/WordPress/gutenberg/issues/20867).

Soean commented:
Wow, this is really great! 🎉 

Short feedback:
- Patterns title is not visible
- Patterns should be searchable
- Inserter tips should be dismissible
- We need drag & drop from the inserter

simison commented:
Would it make sense to call "block patterns" just patterns? The fact they consist of blocks shouldn't matter much to user?

<img width="295" alt="Screenshot 2020-04-17 at 19 14 47" src="https://user-images.githubusercontent.com/87168/79590467-bae73b80-80df-11ea-9c6a-4884611abf20.png">

youknowriad commented:
@simison that's already the case :) and I confirm that I added the logic to remove the tab when you unregister everything.

@Soean good feedback, the patterns title removal was a design choice (discussed a bit above)
we'll work on some follow-ups for the other points, feel free to try a PR for one of these if you're interested too :)

justintadlock commented:
On teams of writers, having the pattern title visible is helpful.  One team member asks, "Which pattern is that you used in X post?"  Another team member can tell them the actual title of the pattern, which can be seen in text and hopefully searchable.  Just a scenario I've already seen come up.

jasmussen commented:
> On teams of writers, having the pattern title visible is helpful. One team member asks, "Which pattern is that you used in X post?" Another team member can tell them the actual title of the pattern, which can be seen in text and hopefully searchable. Just a scenario I've already seen come up.

Totally valid, and easy to pull back. We can revisit this.
